### PR TITLE
Практика 4 – k ближайших соседей

### PR DESCRIPTION
**Pull request by Kolompay**
Выполнено задание по практике №4.

### CODE REVIEW

File: 04_pract_kNN/kNN.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    new = np.zeros((17, 2))
+    new[:16, :2] = sub_X
+    new[16] = obj
+    new = normalize(new, axis=0, norm='max')
+    for i in range(sub_X.shape[0] - 1):
+        sub_X[i] = new[i]
+    obj = new[16]
+    a = [dist(i, obj) for i in sub_X]
+    b = np.argsort(a)
+    b = b[0:k]
```
Comment (fluffystar43) [inline 04_pract_kNN/kNN.py]: Для более понятного анализа кода необходимо указывать названия, отражающие суть происходящего.

**Review by fluffystar43 — CHANGES_REQUESTED** [review]

**Review by fluffystar43 — COMMENTED** [review]
Замечание было исправлено!

**Review by fluffystar43 — APPROVED** [review]

File: 04_pract_kNN/kNN.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    obj = new[16]
+    distance = [dist(i, obj) for i in sub_X]
+    sorting = np.argsort(distance)
+    sortingMas = sorting[0:k]
+    nearest_classes = X[[sortingMas], -1]
+    unique, counts = np.unique(nearest_classes, return_counts = True)
+    object_class = unique[np.argmax(counts)]
+    return object_class
+
+def dist(p1, p2):
```
Comment (stanislavlyalin) [inline 04_pract_kNN/kNN.py]: Лучше переименовать в `euclidean_distance`, чтобы название отражало суть функции.

File: 04_pract_kNN/kNN.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+def k_nearest(X, k, obj):
+    sub_X = X[:, 0: -1]
+    new = np.zeros((17, 2))
+    new[:16, :2] = sub_X
+    new[16] = obj
+    new = normalize(new, axis=0, norm='max')
+    for i in range(sub_X.shape[0] - 1):
+        sub_X[i] = new[i]
+    obj = new[16]
```
Comment (stanislavlyalin) [inline 04_pract_kNN/kNN.py]: Строки 7-13: очень громоздкое решение. К тому же привязка к числу особей в выборке. Функция должна быть универсальной и работать не только для обезьянок, а для любых многомерных данных. Предлагаю исправить так:
1. выполнить стандартизацию sub_X, т.е. для каждого массива-столбца выполнить следующее: из каждого элемента вычесть среднее массива-столбца, а затем поделить на среднеквадратическое отклонение массива-столбца. Можете для этого написать свою небольшую функцию, а можете использовать [StandardScaler из sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html). При этом параметры стандартизации нужно запомнить, т.е. для каждого столбца запомнить среднее и среднеквадратическое отклонение.
2. выполнить стандартизацию объекта `obj`. Из него вычесть поэлементно массив средних значений по столбцам, а затем поделить поэлементно на массив среднеквадратических отклонений.
Текста много, но в коде это выглядит просто. Используйте функции `np.mean` и `np.std` с правильными параметрами `axis`.

**Review by stanislavlyalin — CHANGES_REQUESTED** [review]

**Review by Kolompay — COMMENTED** [review]
Исправил все замечания
